Don't use reserved identifiers memczero and benchmark_verify_t

real-or-random opened the pull request:
As identified in #829 and #833. Fixes #829.

Since we touch this anyway, this commit additionally makes the
identifiers in the benchmark files a little bit more consistent.

This is necessary before we can merge #833. I preferred a separate PR because it makes it easier to see the results of Travis in #833.

elichai commented:
FWIW http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2572.pdf

gmaxwell commented:
@elichai see #833

real-or-random commented:
@elichai @gmaxwell Mind to review? It's trivial.

elichai commented:
Just realized,
are these also UB? :O 
https://github.com/bitcoin-core/secp256k1/blob/ac05f61fcf639a15b5101131561620303e4bd808/src/util.h#L269-L270

elichai commented:
tACK e89278f211a526062745c391d48a7baf782b4b2b

real-or-random commented:
> Just realized,
> are these also UB? :O
> 
> https://github.com/bitcoin-core/secp256k1/blob/ac05f61fcf639a15b5101131561620303e4bd808/src/util.h#L269-L270

Yeah, probably. We use `__` types but we don't declare them, so this fine. But we define types reserved for the standard lib, so I guess this is evil, yes... And this is probably the only case where we may see a real clash in practice in the future, if some compiler defines both `int128_t` and `__int128_t`. (Even then is unlikely to cause issues because it's hopefully the same type.) What we could do is to guard these typedefs with `defined(INT128_MAX)`. Does this make sense? I could add this to the PR.

IMO the entire libary would be more readable if we used typedefs like i32 and u128. But we'd get a huge diff now, and would interrupt all PRs, pretty much like clang-format.

sipa commented:
@real-or-random Replacing things everywhere is invasive, but I think can reasonably change `uint128_t` into `uint128`.

gmaxwell commented:
I had thought the int128 typedefs only ran if they weren't already defined.  I agree clashing here is not great.

Personally I really don't like it when applications make local renames of all the generic standard types:  Then I have to go digging through the code to make sure their custom type is defined to be the thing I think it is and that it's consistently the thing I think it is and isn't something different on some platform or another.  Among other issues you end up with interface uglyness where your custom defined types end up in the interfaces and leak into applications.  It was excusable back before there were standard fixed-size types, but isn't really today, and I think today it mostly still lingers on as copying a style that doesn't make sense anymore.

This is distinct from e.g. making a type for a particular kind of data like a scalar or a field type, because at least that is defined to be fit for a particular purpose.  In most cases using a for-application-type is preferable to using generic types (and, in fact, mandated pretty much universally by MISRA C).

If a generic type needs to be renamed for some compatibility reason-- e.g. there isn't a universally available generic type with the required properties, then sure.

sipa commented:
ACK e89278f211a526062745c391d48a7baf782b4b2b

real-or-random commented:
> I had thought the int128 typedefs only ran if they weren't already defined. 

I added a commit that implements this. I also added `MAX` and `MIN` macros

Note: Currently there's no compiler that provides `(u)int128_t` because that will open a can of worms due to `(u)intmax_t`. See http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2425.pdf .

sipa commented on src/util.h:
```diff
@@ -260,14 +260,19 @@ static SECP256K1_INLINE void secp256k1_int_cmov(int *r, const int *a, int flag)
 # define SECP256K1_WIDEMUL_INT64 1
 #endif
 #if defined(SECP256K1_WIDEMUL_INT128)
+# if !defined(UINT128_MAX) && defined(__SIZEOF_INT128__)
```
This `#if` is missing a corresponding `#endif`.

real-or-random replied:
I hate the preprocessor.

jonasnick reviewed (commented):
Just to spell it out, the assumption is that a compiler that adds int128_t in the future also defines UINT128_MAX? 

Looks good in general. Not sure how useful it is to define constants we don't use.

real-or-random commented:
> Just to spell it out, the assumption is that a compiler that adds int128_t in the future also defines UINT128_MAX?

Yes. And this is implied by the standard as far as I understand it. 

> 
> Looks good in general. Not sure how useful it is to define constants we don't use.

Yeah, one of the reasons why I decided for it is that it increases the probability to have a clash be apparent at compile-time (and fail compilation).

jonasnick reviewed (commented):
ACK 1f4dd0383807bfb7fef884601357b4c629dfb566

sipa commented:
utACK 1f4dd0383807bfb7fef884601357b4c629dfb566
